---
title: Rumen bacterial communities can be acclimated faster to high concentrate diets than currently implemented feedlot programs
author: "Christopher L. Anderson (canderson30@unl.edu)"
output:
  github_document: default
  html_notebook: default
---
This is an updated analysis of [Anderson et al. JAM manuscript]  (http://onlinelibrary.wiley.com/doi/10.1111/jam.13039/full)". The outputs generated through this analysis are identical to those of the [original analysis] (https://github.com/chrisLanderson/old_Anderson_et_al_JAM_Acclimation). Updated analysis was conducted to mostly attempt better coding and reproducible research practices.

Recent studies have demonstrated RAMP, a complete starter feed, to have beneficial effects for animal performance. To understand if RAMP adaptation results in changes in the rumen bacterial community that can potentially affect animal performance, we investigated the dynamics of rumen bacterial community composition in corn‐adapted and RAMP‐adapted cattle.W e compared the bacterial community dynamics in corn and RAMP‐adapted using 16S rRNA gene amplicon sequencing.

```{r, include = FALSE}
install.packages(c("biom", "vegan"), repos = "http://cran.us.r-project.org")
source("http://bioconductor.org/biocLite.R")
biocLite(c("Biostrings", "Heatplus"), suppressUpdates = T)
library(knitr)
library(rmarkdown)
library(vegan)
library(tidyverse)
library(magrittr)
library(stringr)
library(forcats)
library(XML)
library(grid)
library(gplots)
library(biom)
library(RColorBrewer)
library(Biostrings)
```

```{r knitr.settings, include = FALSE}
opts_chunk$set("tidy" = T)
opts_chunk$set("echo" = T)
opts_chunk$set("eval" = T)
opts_chunk$set("warning" = F)
opts_chunk$set("message" = F)
opts_chunk$set("cache" = T)
```

```{r}
source("functions.R")
sessionInfo()
```

## Data Curation

Raw data is provided in the github repository, but needs to be uncompressed. Additionally, here we download the SILVA v119 reference files.

```{r, engine = "bash", results = "hide"}
tar -zxvf rumen_acclimation.tar.gz
curl -LO 'http://www.mothur.org/w/images/2/27/Silva.nr_v119.tgz'
tar -zxvf Silva.nr_v119.tgz
rm -rf Silva.nr_v119.tgz silva.nr_v119.tax Miniconda2-4.2.12-MacOSX-x86_64.sh README*
```

## Demulitplex and Quality Control

The code chunk below demulitplexes the sequencing library using the provided mapping file then trims off the reverse primer.  Subseqeuntly, we trim the seqeunces to a fixed length of 400 basepairs to improve OTU picking downstream. Finally, the sequences are reverse complemented.

```{r, engine="bash", results="hide"}
split_libraries.py -m mapping.txt -f rumen_acclimation.fasta -b hamming_8 -l 0 -L 1000 -M 1 -o demultiplex

truncate_reverse_primer.py -f demultiplex/seqs.fna -o trunc_primer -m mapping.txt -z truncate_only -M 2
  
mothur "#trim.seqs(fasta=trunc_primer/seqs_rev_primer_truncated.fna, minlength=400)"
  
fastx_trimmer -i trunc_primer/seqs_rev_primer_truncated.trim.fasta -l 400 -o qc_trim.fasta
 
mothur "#reverse.seqs(fasta=qc_trim.fasta)"
```

## OTU Picking

Convert the fasta file headers from QIIME format to a format that works with UPARSE to generate the OTU table.

```{r}
seqs <- readDNAStringSet("qc_trim.rc.fasta", format = "fasta")
 
names(seqs) <- str_extract(names(seqs), "R\\d+_\\d+") %>% 
  str_split("_") %>%
  map_chr(~str_c("rumen", .x[[2]], ";barcode=", .x[[1]]))
 
writeXStringSet(seqs, "qc_trim_rc_format.fasta", format = "fasta")
```
 
Use UPARSE suite of tools to pick OTUs.

```{r, engine = "bash"}
gzip -d gold.fasta.gz
mkdir usearch_out

usearch -derep_fulllength qc_trim_rc_format.fasta -sizeout -output usearch_out/derep.fasta

usearch -sortbysize usearch_out/derep.fasta -minsize 2 -output usearch_out/derep_sort.fasta

usearch -cluster_otus usearch_out/derep_sort.fasta -otus usearch_out/otus1.fasta

usearch -uchime_ref usearch_out/otus1.fasta -db gold.fasta -strand plus -nonchimeras usearch_out/otus1_no_chim.fasta

python usearch_python_scripts/fasta_number.py usearch_out/otus1_no_chim.fasta > usearch_out/otus2.fasta

usearch -usearch_global qc_trim_rc_format.fasta -db usearch_out/otus2.fasta -strand plus -id 0.97 -uc usearch_out/otu_map.uc

python usearch_python_scripts/uc2otutab.py usearch_out/otu_map.uc > otu_table.txt
```

## Assign and Summarize Taxonomy

```{r, engine = "bash"}
assign_taxonomy.py -i usearch_out/otus2.fasta -t anaconda/envs/qiime/lib/python2.7/site-packages/qiime_default_reference/gg_13_8_otus/taxonomy/97_otu_taxonomy.txt -r anaconda/envs/qiime/lib/python2.7/site-packages/qiime_default_reference/gg_13_8_otus/rep_set/97_otus.fasta -o assign_taxa -m mothur
```

Add assigned taxa to the OTU table with the column header "taxonomy" and convert the resulting file to biom format.

```{r, engine = 'bash'}
awk 'NR==1; NR > 1 {print $0 | "sort"}' otu_table.txt > otu_table_sort.txt 

sort assign_taxa/otus2_tax_assignments.txt > assign_taxa/otus2_tax_assignments_sort.txt

{ printf '\ttaxonomy\t\t\n'; cat assign_taxa/otus2_tax_assignments_sort.txt ; }  > assign_taxa/otus2_tax_assignments_sort_label.txt

paste otu_table_sort.txt <(cut -f 2 assign_taxa/otus2_tax_assignments_sort_label.txt) > otu_table_tax.txt

rm otu_table_sort.txt

biom convert --table-type "OTU table" -i otu_table_tax.txt -o otu_table_tax.biom --process-obs-metadata taxonomy --to-json
```

## Filter Reads

Two of the samples were collected and sequenced twice. Unsure sure why, but lets remove the duplicate with the lower depth - samples R6 and R19.

```{r, engine='bash'}
printf "R6\nR19" > remove_samples.txt

filter_samples_from_otu_table.py -i otu_table_tax.biom -o otu_table_tax_filter.biom --sample_id_fp remove_samples.txt --negate_sample_id_fp

biom summarize-table -i otu_table_tax_filter.biom -o otu_table_tax_filter_summary.txt
```

Align the sequences using the SILVA reference within mothur and view the alignment summary.

```{r, engine = 'bash', results = 'hide'}
mothur "#align.seqs(fasta=usearch_out/otus2.fasta, reference=silva.nr_v119.align)"

mv usearch_out/otus2.align ./

mothur "#summary.seqs(fasta=otus2.align)"
```

Use R to find the OTUs with poor alignments via the summary file generated above. Decided to remove OTUs that did not end exactly at position 13125 (remember this is the end we sequenced off of...) and started before position 1726.

```{r}
read_tsv("otus2.summary") %>% 
  filter(end == 13125 & start > 1726) %>% 
  select(seqname) %>% 
  write_tsv("remove_otus.txt", col_names = F)
```

Next, remove those OTUs that did not align well and additionally, OTUs with a Cyanobacteria classification. UPARSE should have removed sinlgeton OTUs, but while we are removing OTUs we want to ensure this is the case (-n 2 parameter).

```{r, engine='bash'}
filter_otus_from_otu_table.py -i otu_table_tax_filter.biom -o otu_table_tax_filter2.biom -e remove_otus.txt -n 2 --negate_ids_to_exclude

filter_taxa_from_otu_table.py -i otu_table_tax_filter2.biom -o otu_table_tax_final.biom -n p__Cyanobacteria

biom summarize-table -i otu_table_tax_final.biom -o otu_table_tax_final_summary.txt
```

Summarize taxonomy based on those assignments that go to at least the genera and family levels.

```{r, engine="bash"}
biom convert --table-type "OTU table" -i otu_table_tax_final.biom -o otu_table_tax_final.txt --header-key taxonomy --output-metadata-id "taxonomy" --to-tsv

# summarize by genera
head -n 2 otu_table_tax_final.txt > otu_table_tax_genus.txt

grep "g__[a-zA-Z0-9]" otu_table_tax_final.txt >> otu_table_tax_genus.txt

biom convert --table-type "OTU table" -i otu_table_tax_genus.txt -o otu_table_tax_genus.biom --process-obs-metadata taxonomy --to-json

biom summarize-table -i otu_table_tax_genus.biom -o otu_table_tax_genus_summary.txt

# summarize by family
head -n 2 otu_table_tax_final.txt > otu_table_tax_family.txt

grep "f__[a-zA-Z0-9]" otu_table_tax_final.txt >> otu_table_tax_family.txt 

biom convert --table-type "OTU table" -i otu_table_tax_family.txt -o otu_table_tax_family.biom --process-obs-metadata taxonomy --to-json

biom summarize-table -i otu_table_tax_family.biom -o otu_table_tax_family_summary.txt
```

## Phylogenetic Tree of OTU Representative Sequences

Leaving the OTUs that we removed from the OTU table within the aligned file is fine for downstream analyses. Using that algined file here we generate a phylogenetic tree using the clearcut application in mothur. For this to work, clearcut requires ID lengths greater than ~10 characters. To account for this, we simply add 10 'A's to the front of all sequence names. We then remove the 'A's after the tree is formed.

```{r, engine='bash', results='hide'}
sed -i -e 's/>/>AAAAAAAAAA/g' otus2.align

mothur "#dist.seqs(fasta=otus2.align, output=lt)"

mothur "#clearcut(phylip=otus2.phylip.dist)"

sed -i -e 's/AAAAAAAAAA//g' otus2.phylip.tre
```

## Rarefaction Curves and Alpha Diversity

We wanted to look at the sequencing depth of each sample by monitoring the number of novel OTUs encountered as sample depth is increased. Setup here is for a depth roughly equivalent to least sample seqeunced within a step-up diet in our study so we can visually see full depth to give us an idea if the curves were plateauing. Also, here we compare alpha diversity (observed OTUs and chao1 index) with all samples at the sample depth.

Remember, from QIIME notes: "If the lines for some categories do not extend all the way to the right end of the x-axis, that means that at least one of the samples in that category does not have that many sequences."

```{r, engine='bash'}
multiple_rarefactions.py -i otu_table_tax_final.biom -o alpha_rare -m 10 -x 6600 -s 500 -n 10
 
alpha_diversity.py -i alpha_rare/ -o alpha_rare_otu_chao -m observed_otus,chao1
 
collate_alpha.py -i alpha_rare_otu_chao/ -o alpha_rare_collate
 
make_rarefaction_plots.py -i alpha_rare_collate/ -m mapping.txt -e stderr --generate_average_tables -b Treatment -w -o alpha_rare_collate_avgtable
 
multiple_rarefactions_even_depth.py -i otu_table_tax_final.biom -n 10 -d 2160 -o mult_even
 
alpha_diversity.py -i mult_even/ -o alpha_even -m observed_otus,chao1,goods_coverage 
 
collate_alpha.py -i alpha_even -o alpha_even_collate

alpha_diversity.py -i otu_table_tax_final.biom -o goods.txt -m goods_coverage
```

```{r, fig.width=10, fig.height=10}
rare_html <- readHTMLTable("alpha_rare_collate_avgtable/rarefaction_plots.html")
rare_html$rare_data[rare_html$rare_data == "nan"] <- NA
colnames(rare_html$rare_data) <- c("treatment", "seqs", "chao_avg", "chao_err", "otu_avg", "otu_err")
rare_df <- as_tibble(rare_html$rare_data)
rare_df <- rare_df %>% 
  mutate_at(vars(-treatment), as.character) %>% 
  mutate_at(vars(-treatment), as.double) %>% 
  mutate(diet = if_else(str_detect(treatment, "^C"), "Corn", "RAMP")) %>%
  mutate(plot_color = case_when(
    str_detect(rare_df$treatment, ("[CR]1")) ~ "#FF0000",
    str_detect(rare_df$treatment, ("[CR]2")) ~ "#BF003F",
    str_detect(rare_df$treatment, ("[CR]3")) ~ "#7F007F",
    str_detect(rare_df$treatment, ("[CR]4")) ~ "#3F00BF",
    str_detect(rare_df$treatment, ("[CR]F")) ~ "#0000FF"
  )) %>% 
  group_by(diet)

plot_color_vect <- rare_df$plot_color
names(plot_color_vect) <- rare_df$treatment

rare_otu_plot <- ggplot(rare_df, aes(x = seqs, y = otu_avg, color = treatment, group = treatment, ymin = otu_avg - otu_err, ymax = otu_avg + otu_err)) +
  facet_grid(~diet) +
  geom_line(position = position_dodge(width = 275)) + 
  geom_pointrange(position = position_dodge(width = 275), size = 0.3) +
  scale_color_manual(values = plot_color_vect) +
  labs(x = "Sequences per Sample", y = "Mean Observed OTUs") +
  theme(legend.title = element_blank(), 
        text = element_text(size = 6))

rare_chao_plot <- ggplot(rare_df, aes(x = seqs, y = chao_avg, color = treatment, group = treatment, ymin = chao_avg - chao_err, ymax = chao_avg + chao_err)) +
  facet_grid(~diet) +
  geom_line(position = position_dodge(width = 275)) + 
  geom_pointrange(position = position_dodge(width = 275), size = 0.3) +
  scale_color_manual(values = plot_color_vect) +
  labs(x = "Sequences per Sample", y = "Mean Observed OTUs") +
  theme(legend.title = element_blank(), 
        text = element_text(size = 6))

alpha_col_names <- c("CF_332", "R3_259", "R2_343", "R4_343", "R4_259", "C3_346", "C4_332", "R1_222", "CF_346", "RF_343", "RF_222", "C3_332", "R2_222", "R1_343", "R3_343", "C1_346", "C2_332", "R2_259", "R3_222", "C1_332", "C2_346", "R4_222", "RF_259", "R1_259", "C4_346")

alpha_chao <- read_tsv("alpha_even_collate/chao1.txt")
alpha_chao <- select(alpha_chao, starts_with("R"))
names(alpha_chao) <- alpha_col_names
alpha_chao_t <- enframe(as.list(alpha_chao))
alpha_chao_plot <- alpha_chao_t %>% 
  mutate(means = map_dbl(value, mean)) %>% 
  mutate(sds = map_dbl(value, sd)) %>% 
  mutate(diet = if_else(str_detect(name, "^C"), "Corn", "RAMP")) %>% 
  mutate(treatment = str_sub(name, 1, 2)) %>% 
  ggplot(., aes(x = treatment, y = means)) +
    facet_wrap(~diet) +
    geom_point(size = 2) + 
    labs(x = "", y = "Mean Chao1 Index") +
    theme(legend.title = element_blank(),text = element_text(size = 6))

alpha_otu <- read_tsv("alpha_even_collate/observed_otus.txt")
alpha_otu <- select(alpha_otu, starts_with("R"))
names(alpha_otu) <- alpha_col_names
alpha_otu_t <- enframe(as.list(alpha_otu))
alpha_otu_plot <- alpha_otu_t %>% 
  mutate(means = map_dbl(value, mean)) %>% 
  mutate(sds = map_dbl(value, sd)) %>% 
  mutate(diet = if_else(str_detect(name, "^C"), "Corn", "RAMP")) %>% 
  mutate(treatment = str_sub(name, 1, 2)) %>% 
  ggplot(., aes(x = treatment, y = means)) +
    facet_wrap(~diet) +
    geom_point(size = 2) + 
    labs(x = "", y = "Mean Observed OTUs") +
    theme(legend.title = element_blank(),text = element_text(size = 6))

multiplot(rare_otu_plot, rare_chao_plot, alpha_otu_plot, alpha_chao_plot, cols = 2)

pdf("FigureS2.pdf", height = 12, width = 12)
multiplot(rare_otu_plot, rare_chao_plot, alpha_otu_plot, alpha_chao_plot, cols = 2)
dev.off()

```

## Beta Diversity

Investigate the influence of Corn and RAMP based treatments on beta diversity independently.

```{r, engine='bash'}
biom summarize-table -i otu_table_tax_final.biom -o otu_table_tax_final_summary.txt

biom convert -i otu_table_tax_final.biom -o otu_table_tax_final.json.biom --to-json --table-type="OTU table"

beta_diversity_through_plots.py -i otu_table_tax_final.biom -o beta_div -t otus2.phylip.tre -m mapping.txt -p qiime_parameters_working.txt -e 2160

split_otu_table.py -i otu_table_tax_final.biom -o split_total -m mapping.txt -f Diet
 
biom summarize-table -i split_total/otu_table_tax_final__Diet_Corn__.biom -o split_total/corn_summarize.txt

biom summarize-table -i split_total/otu_table_tax_final__Diet_RAMP__.biom -o split_total/ramp_summarize.txt

beta_diversity_through_plots.py -i split_total/otu_table_tax_final__Diet_Corn__.biom -o corn_beta_div -t otus2.phylip.tre -m mapping.txt -p qiime_parameters_working.txt -e 2160

beta_diversity_through_plots.py -i split_total/otu_table_tax_final__Diet_RAMP__.biom -o ramp_beta_div -t otus2.phylip.tre -m mapping.txt -p qiime_parameters_working.txt -e 2959
```

We use PERMANOVA with the unweighted UNIFRAC distance matrix to identify changes in microbial community structure between step-up treatment groups in those animals fed RAMP and the corn-based diet. Additioanlly, we check that the data does not violate the assumptions of PERMANOVA. However, because there is no variance within the corn-based treatment group due to low sample numbers (that is, within a group there is only one distance value because *n* = 2, i.e. C1-C1 distance, C2-C2 distance, etc.), the results testing a violation of the assumptions of PERMANOVA are unreliable.

```{r}
set.seed(11)
corn_unweighted <- read_tsv("corn_beta_div/unweighted_unifrac_dm.txt")
corn_unweighted_dist <- corn_unweighted %>% 
  select(-X1) %>% 
  as.dist(.)

ramp_unweighted <- read_tsv("ramp_beta_div/unweighted_unifrac_dm.txt")
ramp_unweighted_dist <- ramp_unweighted %>% 
  select(-X1) %>% 
  as.dist(.)

dist_list <- list(corn_unweighted_dist, ramp_unweighted_dist)

mapping <- read_tsv("mapping.txt")
names(mapping)[1] <- "SampleID"
mapping <- mapping %>% 
  filter(SampleID != "R6", SampleID != "R19") %>% 
  mutate_at(vars(AnimalID, Treatment), as.factor) %>% 
  group_by(Diet) %>% 
  nest() %>% 
  arrange(Diet) %>%
  mutate(dists = dist_list) %>% 
  mutate(permanova = map2(dists, data, ~adonis(.x ~ Treatment + AnimalID, permutations = 999, data = .y))) %>% 
  mutate(anova_treat = map2(dists, data, ~anova(betadisper(.x, .y$Treatment)))) %>% 
  mutate(anova_animal = map2(dists, data, ~anova(betadisper(.x, .y$AnimalID)))) %>%
  mutate(permu_treat = map2(dists, data, ~permutest(betadisper(.x, .y$Treatment)))) %>% 
  mutate(permu_animal = map2(dists, data, ~permutest(betadisper(.x, .y$AnimalID))))
```

Results of statistical tests were saved in the data frame...print them to screen.

```{r}
(corn_perm <- mapping %>% 
  filter(Diet == "Corn") %>% 
  select(permanova)
)[[1]]
```

```{r}
(corn_anova_treat <- mapping %>% 
  filter(Diet == "Corn") %>% 
  select(anova_treat)
)[[1]]
```

```{r}
(corn_anova_animal <- mapping %>% 
  filter(Diet == "Corn") %>% 
  select(anova_animal)
)[[1]]
```

```{r}
(corn_permu_treat <- mapping %>% 
  filter(Diet == "Corn") %>% 
  select(permu_treat)
)[[1]]
```

```{r}
(corn_permu_animal <- mapping %>% 
  filter(Diet == "Corn") %>% 
  select(permu_animal)
)[[1]]
```

```{r}
(ramp_perm <- mapping %>% 
  filter(Diet == "RAMP") %>% 
  select(permanova)
)[[1]]
```

```{r}
(ramp_anova_treat <- mapping %>% 
  filter(Diet == "RAMP") %>% 
  select(anova_treat)
)[[1]]
```

```{r}
(ramp_anova_animal <- mapping %>% 
  filter(Diet == "RAMP") %>% 
  select(anova_animal)
)[[1]]
```

```{r}
(ramp_permu_treat <- mapping %>% 
  filter(Diet == "RAMP") %>% 
  select(permu_treat)
)[[1]]
```

```{r}
(ramp_permu_animal <- mapping %>% 
  filter(Diet == "RAMP") %>% 
  select(permu_animal)
)[[1]]
```

We used the Mann-Whitney U Test as a post-hoc method to compare the microbial communities at stages within an adaptation program. To do this, we compared the pairwise distances relative to step 1 of adaptation. For instance, to test for a difference in community structure between step 2 and step 3, we compared all the pairwise distances from step 1 samples to step 2 samples (S1-S2) to the distance from step 1 samples to step 3 samples (S1-S3).

```{r}
corn_unweighted_tidy <- corn_unweighted %>%
  gather(R10:R9, key = S2, value = dist) %>% 
  dplyr::rename(S1 = X1) %>% 
  mutate(unique_key = paste0(pmin(S1, S2), pmax(S1, S2), sep = "")) %>% 
  distinct(unique_key, .keep_all = T) %>% 
  select(-unique_key)

mapping <- read_tsv("mapping.txt")
names(mapping)[1] <- "SampleID"

corn_unweighted_tidy <- left_join(corn_unweighted_tidy, mapping, by = c("S1" = "SampleID")) %>% 
  select(S1, S2, dist, Treatment) %>% 
  dplyr::rename(S1_treatment = Treatment) %>% 
  left_join(., mapping, by = c("S2" = "SampleID")) %>% 
  dplyr::rename(S2_treatment = Treatment) %>% 
  mutate(dist_treatment = map2_chr(S1_treatment, S2_treatment, collapse_treatment)) %>% 
  filter(dist != 0) %>% 
  select(dist, dist_treatment) %>% 
  group_by(dist_treatment) %>% 
  nest()

corn_unweighted_tidy %>% 
  filter((dist_treatment == "C1-C1") | (dist_treatment == "C1-C2")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
corn_unweighted_tidy %>% 
  filter((dist_treatment == "C1-C2") | (dist_treatment == "C1-C3")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
corn_unweighted_tidy %>% 
  filter((dist_treatment == "C1-C3") | (dist_treatment == "C1-C4")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
corn_unweighted_tidy %>% 
  filter((dist_treatment == "C1-C4") | (dist_treatment == "C1-CF")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
ramp_unweighted_tidy <- ramp_unweighted %>%
  gather(R11:R8, key = S2, value = dist) %>% 
  dplyr::rename(S1 = X1) %>% 
  mutate(unique_key = paste0(pmin(S1, S2), pmax(S1, S2), sep = "")) %>% 
  distinct(unique_key, .keep_all = T) %>% 
  select(-unique_key)

ramp_unweighted_tidy <- left_join(ramp_unweighted_tidy, mapping, by = c("S1" = "SampleID")) %>% 
  select(S1, S2, dist, Treatment) %>% 
  dplyr::rename(S1_treatment = Treatment) %>% 
  left_join(., mapping, by = c("S2" = "SampleID")) %>% 
  dplyr::rename(S2_treatment = Treatment) %>% 
  mutate(dist_treatment = map2_chr(S1_treatment, S2_treatment, collapse_treatment)) %>% 
  filter(dist != 0) %>% 
  select(dist, dist_treatment) %>% 
  group_by(dist_treatment) %>% 
  nest()

ramp_unweighted_tidy %>% 
  filter((dist_treatment == "R1-R1") | (dist_treatment == "R1-R2")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
ramp_unweighted_tidy %>% 
  filter((dist_treatment == "R1-R2") | (dist_treatment == "R1-R3")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
ramp_unweighted_tidy %>% 
  filter((dist_treatment == "R1-R3") | (dist_treatment == "R1-R4")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

```{r}
ramp_unweighted_tidy %>% 
  filter((dist_treatment == "R1-R4") | (dist_treatment == "R1-RF")) %$%
  wilcox.test(data[[1]] %$% dist, data[[2]] %$% dist)
```

## Shared OTUs and Sequences

First, make venn diagrams dislaying the distribution of OTUs between steup up diets within the RAMP and corn-based treatment.

```{r, engine='bash'}
collapse_samples.py -b otu_table_tax_final.biom -m mapping.txt --output_biom_fp otu_table_collapse_samples.biom --output_mapping_fp mapping_collapse.txt --collapse_fields Treatment --collapse_mode sum

biom convert -i otu_table_collapse_samples.biom -o otu_table_collapse_samples.json.biom --to-json --table-type="OTU table"

split_otu_table.py -i otu_table_collapse_samples.biom -m mapping_collapse.txt -f Diet -o split_collapse_samples

biom convert -i split_collapse_samples/otu_table_collapse_samples__Diet_Corn__.biom -o split_collapse_samples/otu_table_collapse_samples__Diet_Corn__.json.biom --to-json --table-type="OTU table"

biom convert -i split_collapse_samples/otu_table_collapse_samples__Diet_RAMP__.biom -o split_collapse_samples/otu_table_collapse_samples__Diet_RAMP__.json.biom --to-json --table-type="OTU table"
```

```{r, fig.width=6, fig.height=6}
corn_biom <- read_biom("split_collapse_samples/otu_table_collapse_samples__Diet_Corn__.json.biom")

corn_df <- as_tibble(as(biom_data(corn_biom), "matrix"))
corn_boolean_df <- as_tibble(corn_df > 0) %>% 
  select(C1, C2, C3, C4, CF)

corn_venn <- venn(corn_boolean_df)

pdf("Figure4_panelA.pdf", height = 6, width = 6)
corn_venn <- venn(corn_boolean_df)
dev.off()

ramp_biom <- read_biom("split_collapse_samples/otu_table_collapse_samples__Diet_RAMP__.json.biom")

ramp_df <- as_tibble(as(biom_data(ramp_biom), "matrix"))
ramp_boolean_df <- as_tibble(ramp_df > 0) %>% 
  select(R1, R2, R3, R4, RF)

ramp_venn <- venn(ramp_boolean_df)

pdf("Figure4_panelB.pdf", height = 6, width = 6)
ramp_venn <- venn(ramp_boolean_df)
dev.off()
```

Get shared number of OTUs and sequences per sample.

```{r, engine='bash'}
shared_phylotypes.py -i otu_table_collapse_samples.biom -o collapse_samples_shared_otus.txt
```

Write shared number of OTUs between treatments to file.

```{r}
share_otu <- read_tsv("collapse_samples_shared_otus.txt")

share_otu_tidy <- share_otu %>% 
  gather(R4:C4, key = "S2", value = "OTUs") %>% 
  dplyr::rename(S1 = X1) %>% 
  mutate(unique_key = paste0(pmin(S1, S2), pmax(S1, S2), sep = "")) %>% 
  distinct(unique_key, .keep_all = T) %>% 
  select(-unique_key) %>% 
  filter(S1 != S2) %>% 
  mutate(max_share = map2_int(S1, S2, max_shared_otu)) %>% 
  mutate(per_share = (OTUs / max_share) * 100) %>% 
  select(S1, S2, OTUs, per_share) %>% 
  write_tsv("Table_S4.txt", col_names = F)
```

Write shared number of sequences between treatments to file.

```{r}
collapse_biom <- read_biom("otu_table_collapse_samples.json.biom")
collapse_df <- as_data_frame(as(biom_data(collapse_biom), "matrix"))
collapse_tib <- as_tibble(rownames_to_column(collapse_df))
collapse_tidy <- collapse_tib %>% 
  dplyr::rename(otu = rowname) %>% 
  gather(R4:C4, key = "sample", value = "abundance") %>% 
  filter(abundance > 0) %>% 
  group_by(sample) %>% 
  nest()

combn_output <- combn(unique(collapse_tidy$sample), 2, simplify = FALSE, FUN = seq_shared)
```

## OTU Heatmap

```{r}
otu_table_biom <- read_biom("otu_table_tax_final.json.biom")
otu_table <- as.data.frame(as(biom_data(otu_table_biom), "matrix"))
row.names(otu_table) <- otu_table$OTU.ID
colnames(otu_table) <- c("CF_332", "R3_259", "R2_343", "R4_343", "R4_259", "C3_346", "C4_332", "R1_222", "CF_346", "RF_343", "RF_222", "C3_332", "R2_222", "R1_343", "R3_343", "C1_346", "C2_332", "R2_259", "R3_222", "C1_332", "C2_346", "R4_222", "RF_259", "R1_259", "C4_346")
otu_table_trans <- as.data.frame(t(otu_table))
otu_table_rel <- otu_table_trans/rowSums(otu_table_trans)
scalewhiteblack <- colorRampPalette(c("white", "black"), space = "rgb")(100)
otu_dist <- vegdist(otu_table_rel, method = "bray")
row.clus <- hclust(otu_dist, "aver")
maxab <- apply(otu_table_rel, 2, max)
n1 <- names(which(maxab < 0.01))
otu_table_rel2 <- otu_table_rel[, -which(names(otu_table_rel) %in% n1)]
otu_dist_col <- vegdist(t(otu_table_rel2), method = "bray")
col.clus <- hclust(otu_dist_col, "aver")
png("Figure2.png", units = 'in', height = 8, width = 8, res = 600)
heatmap.2(as.matrix(otu_table_rel2), Rowv = as.dendrogram(row.clus), Colv = as.dendrogram(col.clus), col = scalewhiteblack, margins = c(2, 6), trace = "none", density.info = "none", labCol = "", xlab = "OTUs", ylab = "Samples", main = "", lhei = c(2, 8))
dev.off()
```

![](Figure2.png)

## Pairwise Comparisons and Principle Component Analysis

```{r, fig.width=10, fig.height=4}
corn_distance_plot <- corn_unweighted_tidy %>% 
  filter(str_detect(dist_treatment, "^C1-")) %>% 
  unnest() %>% 
  ggplot(., aes(x = dist_treatment, y = dist)) + 
    geom_boxplot() +
    geom_point(position = position_jitter(width = 0.15)) +
    labs(x = "", y = "Unweighted UniFrac Distance\n") +
    guides(fill = FALSE) +
    theme(axis.text.x = element_text(colour = "black"),  
          axis.title.y = element_text(size = 14), 
          axis.ticks = element_blank())

ramp_distance_plot <- ramp_unweighted_tidy %>% 
  filter(str_detect(dist_treatment, "^R1-")) %>% 
  unnest() %>% 
  ggplot(., aes(x = dist_treatment, y = dist)) + 
    geom_boxplot() +
    geom_point(position = position_jitter(width = 0.15)) +
    labs(x = "", y = "Unweighted UniFrac Distance\n") +
    guides(fill = FALSE) +
    theme(axis.text.x = element_text(colour = "black"),  
          axis.title.y = element_text(size = 14), 
          axis.ticks = element_blank())

shape_plot <- c("R1" = 15, "R2" = 16, "R3" = 17, "R4" = 18, "RF" = 8, "C1" = 15, "C2" = 16, "C3" = 17, "C4" = 18, "CF" = 8)
color_plot <- c("R1" = "red", "R2" = "red", "R3" = "red", "R4" = "red", "RF" = "red", "C1" = "blue", "C2" = "blue", "C3" = "blue", "C4" = "blue", "CF" = "blue")
con <- file("beta_div/unweighted_unifrac_pc.txt")
pc <- read_tsv(con, skip = 9, n_max = 25, col_names = F)
pc_plot <- pc %>% 
  rename(SampleID = X1, PC1 = X2, PC2 = X3) %>% 
  left_join(., mapping, by = "SampleID") %>% 
  select(SampleID, PC1, PC2, Treatment) %>% 
  ggplot(., aes(PC1, PC2)) + 
    geom_point(aes(size=4, color = factor(Treatment), shape = factor(Treatment))) +    
    xlab("PC1 (21.1%)") + ylab("PC2 (10.0%)") +
    guides(size = FALSE) + 
    scale_shape_manual(name = "", values = shape_plot) +
    scale_colour_manual(name = "", values = color_plot) +
    labs(fill = "")

multiplot(pc_plot, corn_distance_plot, ramp_distance_plot, cols = 3)

pdf("Figure1.pdf", height = 6, width = 16)
multiplot(pc_plot, corn_distance_plot, ramp_distance_plot, cols = 3)
dev.off()
```

## OTU-Level Statistics and Comparisons

Here, we use LEFSE to identify OTUs that differed between treatments at potential break points in microbial community structure found above (i.e. from steps C3 to C4, R1 to R2, and R2 to R3). First, we need to format a file suitable for input into LEFSE.

```{r, engine='bash'}
# remove potential singleton OTUs
filter_otus_from_otu_table.py -i split_total/otu_table_tax_final__Diet_Corn__.biom -n 1 -o split_total/otu_table_tax_final__Diet_Corn__filter.biom

filter_otus_from_otu_table.py -i split_total/otu_table_tax_final__Diet_RAMP__.biom -n 1 -o split_total/otu_table_tax_final__Diet_RAMP__filter.biom
```

```{r}
corn_biom <- read_biom("split_total/otu_table_tax_final__Diet_Corn__filter.biom")
corn_table <- as_data_frame(as(biom_data(corn_biom), "matrix"))
corn_rel <- sweep(corn_table, 2, colSums(corn_table), FUN = "/")
rownames(corn_rel) <- paste("OTU", rownames(corn_rel), sep = "")
corn_rel2 <- corn_rel[0, ]
corn_rel2[nrow(corn_rel2) + 1, ] <- c("break2", "break1", "break2", "break2", "break1", "break1", "break1", "break1", "break1", "break2")
corn_rel2[nrow(corn_rel2) + 1, ] <- c("A332", "A346", "A332", "A346", "A332", "A346", "A332", "A332", "A346", "A346")
row.names(corn_rel2) <- c("break", "animal")
corn_rel_merge <- rbind(corn_rel2, corn_rel)
write.table(corn_rel_merge, sep = "\t", file = "split_total/otu_table_tax_final__Diet_Corn__filter_relative.txt", row.names = TRUE, col.names = FALSE, quote = FALSE)

ramp_biom <- read_biom("split_total/otu_table_tax_final__Diet_RAMP__filter.biom")
ramp_table <- as_data_frame(as(biom_data(ramp_biom), "matrix"))
ramp_rel <- sweep(ramp_table, 2, colSums(ramp_table), FUN="/")
rownames(ramp_rel) <- paste("OTU", rownames(ramp_rel), sep="")
ramp_rel2 <- ramp_rel[0,]
ramp_rel2[nrow(ramp_rel2) + 1, ] <- c( "break3", "break2", "break3", "break3", "break1", "break3", "break3", "break2", "break1", "break3", "break2", "break3", "break2", "break3", "break1")
ramp_rel2[nrow(ramp_rel2) + 1, ] <- c( "A259", "A343", "A343", "A259", "A222", "A343", "A222", "A222", "A343", "A343", "A259", "A222", "A222", "A259", "A259")
row.names(ramp_rel2) <- c("break","animal")
ramp_rel_merge <- rbind(ramp_rel2,ramp_rel)
ramp_rel_merge_1 <- subset(ramp_rel_merge, select = c(R18, R24, R8, R12, R23, R5))
ramp_rel_merge_2 <- subset(ramp_rel_merge, select=c(R12, R23, R5, R11, R13, R14, R20, R21, R23, R3, R4, R7))
write.table(ramp_rel_merge_1, sep="\t", file="split_total/otu_table_tax_final__Diet_RAMP__filter_b12_relative.txt", row.names=TRUE, col.names=FALSE, quote = FALSE)
write.table(ramp_rel_merge_2, sep="\t", file="split_total/otu_table_tax_final__Diet_RAMP__filter_b23_relative.txt", row.names=TRUE, col.names=FALSE, quote = FALSE)
```

Run LEFSE.

```{r, engine='bash'}
wget https://bitbucket.org/nsegata/lefse/get/default.zip -O lefse.zip
unzip lefse.zip
mv nsegata* lefse

python lefse/format_input.py split_total/otu_table_tax_final__Diet_Corn__filter_relative.txt corn_lefse_format.txt -c 1 -u 2 -o 1000000 
python lefse/run_lefse.py corn_lefse_format.txt corn_lefse_result.txt

python lefse/format_input.py split_total/otu_table_tax_final__Diet_RAMP__filter_b12_relative.txt ramp_12_lefse_format.txt -c 1 -u 2 -o 1000000 
python lefse/run_lefse.py ramp_12_lefse_format.txt ramp_12_lefse_result.txt

python lefse/format_input.py split_total/otu_table_tax_final__Diet_RAMP__filter_b23_relative.txt ramp_23_lefse_format.txt -c 1 -u 2 -o 1000000 
python lefse/run_lefse.py ramp_23_lefse_format.txt ramp_23_lefse_result.txt
```

```{r}
corn_lefse <- read_tsv("corn_lefse_result.txt", col_names = F)
corn_lefse %>% 
  filter(X4 != "NA") %>% 
  mutate(X1 = str_replace(X1, "OTU", "")) %$% 
  write(X1, "corn_lefse_otus.txt")

ramp_12_lefse <- read_tsv("ramp_12_lefse_result.txt", col_names = F)
ramp_12_lefse %>% 
  filter(X4 != "NA") %>% 
  mutate(X1 = str_replace(X1, "OTU", "")) %$% 
  write(X1, "ramp_12_lefse_otus.txt")

ramp_23_lefse <- read_tsv("ramp_23_lefse_result.txt", col_names = F)
ramp_23_lefse %>% 
  filter(X4 != "NA") %>% 
  mutate(X1 = str_replace(X1, "OTU", "")) %$% 
  write(X1, "ramp_23_lefse_otus.txt")

```

```{r, engine='bash'}
filter_otus_from_otu_table.py -i split_total/otu_table_tax_final__Diet_Corn__filter.biom -o corn_lefse.biom -e corn_lefse_otus.txt --negate_ids_to_exclude
biom convert -i corn_lefse.biom -o corn_lefse.txt --table-type="OTU table" --to-tsv --header-key taxonomy

filter_otus_from_otu_table.py -i split_total/otu_table_tax_final__Diet_RAMP__filter.biom -o ramp_12_lefse.biom -e ramp_12_lefse_otus.txt --negate_ids_to_exclude
biom convert -i ramp_12_lefse.biom -o  ramp_12_lefse.txt --table-type="OTU table" --to-tsv --header-key taxonomy

filter_otus_from_otu_table.py -i split_total/otu_table_tax_final__Diet_RAMP__filter.biom -o ramp_23_lefse.biom -e ramp_23_lefse_otus.txt --negate_ids_to_exclude
biom convert -i ramp_23_lefse.biom -o  ramp_23_lefse.txt --table-type="OTU table" --to-tsv --header-key taxonomy

biom convert -i split_total/otu_table_tax_final__Diet_Corn__filter.biom -o split_total/otu_table_tax_final__Diet_Corn__filter.txt --table-type "OTU table" --to-tsv --header-key taxonomy --output-metadata-id "taxonomy"

biom convert -i split_total/otu_table_tax_final__Diet_RAMP__filter.biom -o split_total/otu_table_tax_final__Diet_RAMP__filter.txt --table-type "OTU table" --to-tsv --header-key taxonomy --output-metadata-id "taxonomy"
```

One shift in the microbial community was identified in the corn-adaptation program. Here, we plot the heatmap for OTUs identified as having a significantly different abundance around this shift. The resulting heatmap only displays OTUs with a maximum relative abundance >1% and are sorted by LDA score.

Corn-based treatment first.

```{r}
corn_otu <- read_tsv("corn_lefse.txt", skip = 1)
names(corn_otu)[1] <- "OTUID"
corn_fams <- corn_otu %>% 
  mutate(family = str_extract(taxonomy, "(?<=f__)[\\[\\]\\w\\d]+")) %>% 
  mutate(family = str_replace_all(family, "[\\[\\]]", "")) %>% 
  mutate(family = str_replace_na(family, replacement = "No Assigned Family")) %>% 
  select(OTUID, family)

corn_lefse <- read_tsv("corn_lefse_result.txt", col_names = F)
corn_lefse <- corn_lefse %>% 
  dplyr::rename(OTUID = X1, raw = X2, breaks = X3, LDA = X4, pval = X5) %>% 
  mutate(OTUID = as.integer(str_replace(OTUID, "OTU", ""))) %>% 
  left_join(corn_fams, ., by = "OTUID") %>% 
  arrange(breaks, desc(LDA)) %>% 
  write_tsv("Table_S2-1.txt")

corn_table <- read_tsv("split_total/otu_table_tax_final__Diet_Corn__filter.txt", skip = 1)
names(corn_table)[1] <- "OTUID"
  
corn_otu_lefse <- corn_table %>% 
  rename(CF_332 = R10, C3_346 = R15, C4_332 = R16, CF_346 = R2, C3_332 = R22, C1_346 = R36, C2_332 = R37, C1_332 = R40, C2_346 = R41, C4_346 = R9) %>%
  select(OTUID, C1_346, C1_332, C2_346, C2_332, C3_346, C3_332, C4_346, C4_332, CF_346, CF_332)  %>% 
  mutate_each(funs(normalize_table), -OTUID) %>% 
  gather(C1_346:CF_332, key = sample, value = abundance) %>% 
  group_by(OTUID) %>% 
  filter(any(abundance > 0.01)) %>% 
  ungroup() %>% 
  spread(key = sample, value = abundance) %>% 
  mutate(OTUID = as.integer(OTUID)) %>% 
  inner_join(corn_lefse, by = "OTUID") %>% 
  arrange(breaks, desc(LDA))

scalewhiteblack <- colorRampPalette(c("white", "black"), space = "rgb")(100)
lmat = rbind(c(4,0,0),c(2,1,3))
lwid = c(0.60,1.9,0.3)
lhei = c(0.3,1.5)

pdf("Figure3_panelA.pdf", width = 12, height = 9)
corn_otu_lefse %>% 
  select(starts_with("C", ignore.case = F)) %>% 
  t() %>% 
  heatmap.2(Rowv = F, Colv = F, col = scalewhiteblack, margins = c(13, 9.5), trace = "none", density.info = "none", xlab = "", ylab = "", main = "", srtCol = 67.5, cexCol = 1.3, cexRow = 2.0, lmat = lmat, lwid = lwid, lhei = lhei, labCol = corn_otu_lefse$family)
dev.off()
```


Same thing, but plotting differentially abundant OTUs around the two shift in microbial community identified in the RAMP adaptation treatment.

```{r}
# first microbial shift between steps 1 and 2
ramp_12_otu <- read_tsv("ramp_12_lefse.txt", skip = 1)
names(ramp_12_otu)[1] <- "OTUID"
ramp_12_fams <- ramp_12_otu %>% 
  mutate(family = str_extract(taxonomy, "(?<=f__)[\\[\\]\\w\\d]+")) %>% 
  mutate(family = str_replace_all(family, "[\\[\\]]", "")) %>% 
  mutate(family = str_replace_na(family, replacement = "No Assigned Family")) %>% 
  select(OTUID, family)

ramp_12_lefse <- read_tsv("ramp_12_lefse_result.txt", col_names = F)
ramp_12_lefse <- ramp_12_lefse %>% 
  dplyr::rename(OTUID = X1, raw = X2, breaks = X3, LDA = X4, pval = X5) %>% 
  mutate(OTUID = as.integer(str_replace(OTUID, "OTU", ""))) %>% 
  left_join(ramp_12_fams, ., by = "OTUID") %>% 
  arrange(breaks, desc(LDA)) %>% 
  write_tsv("Table_S2-2.txt")

ramp_table <- read_tsv("split_total/otu_table_tax_final__Diet_RAMP__filter.txt", skip = 1)
names(ramp_table)[1] <- "OTUID"
ramp_12_otu_lefse <- ramp_table %>% 
  rename(R3_259 = R11, R2_343 = R12, R4_343 = R13, R4_259 = R14, R1_222 = R18, RF_343 = R20, RF_222 = R21, R2_222 = R23, R1_343 = R24, R3_343 = R3, R2_259 = R38, R3_222 = R4, R4_222 = R5, RF_259 = R7, R1_259 = R8) %>%
  select(OTUID, R1_222, R1_259, R1_343, R2_222, R2_259, R2_343)  %>% 
  mutate_each(funs(normalize_table), -OTUID) %>% 
  gather(R1_222:R2_343, key = sample, value = abundance) %>% 
  group_by(OTUID) %>% 
  filter(any(abundance > 0.01)) %>% 
  ungroup() %>% 
  spread(key = sample, value = abundance) %>% 
  mutate(OTUID = as.integer(OTUID)) %>% 
  inner_join(ramp_12_lefse, by = "OTUID") %>% 
  arrange(breaks, desc(LDA))

scalewhiteblack <- colorRampPalette(c("white", "black"), space = "rgb")(100)
lmat = rbind(c(4,0,0),c(2,1,3))
lwid = c(0.60,1.9,0.3)
lhei = c(0.3,1.5)

pdf("Figure3_panelB.pdf", width = 12, height = 9)
ramp_12_otu_lefse %>% 
  select(starts_with("R", ignore.case = F)) %>% 
  t() %>% 
  heatmap.2(Rowv = F, Colv = F, col = scalewhiteblack, margins = c(13, 9.5), trace = "none", density.info = "none", xlab = "", ylab = "", main = "", srtCol = 67.5, cexCol = 1.3, cexRow = 2.0, lmat = lmat, lwid = lwid, lhei = lhei, labCol = ramp_12_otu_lefse$family)
dev.off()

  
# second microbial shift between steps 2 and 3/4/F
ramp_23_otu <- read_tsv("ramp_23_lefse.txt", skip = 1)
names(ramp_23_otu)[1] <- "OTUID"
ramp_23_fams <- ramp_23_otu %>% 
  mutate(family = str_extract(taxonomy, "(?<=f__)[\\[\\]\\w\\d]+")) %>% 
  mutate(family = str_replace_all(family, "[\\[\\]]", "")) %>% 
  mutate(family = str_replace_na(family, replacement = "No Assigned Family")) %>% 
  select(OTUID, family)

ramp_23_lefse <- read_tsv("ramp_23_lefse_result.txt", col_names = F)
ramp_23_lefse <- ramp_23_lefse %>% 
  dplyr::rename(OTUID = X1, raw = X2, breaks = X3, LDA = X4, pval = X5) %>% 
  mutate(OTUID = as.integer(str_replace(OTUID, "OTU", ""))) %>% 
  left_join(ramp_23_fams, ., by = "OTUID") %>% 
  arrange(breaks, desc(LDA)) %>% 
  write_tsv("Table_S2-3.txt")

ramp_23_otu_lefse <- ramp_table %>% 
  rename(R3_259 = R11, R2_343 = R12, R4_343 = R13, R4_259 = R14, R1_222 = R18, RF_343 = R20, RF_222 = R21, R2_222 = R23, R1_343 = R24, R3_343 = R3, R2_259 = R38, R3_222 = R4, R4_222 = R5, RF_259 = R7, R1_259 = R8) %>%
  select(OTUID, R2_222, R2_259, R2_343, R3_222, R3_259, R3_343, R4_222, R4_259, R4_343, RF_222, RF_259, RF_343)  %>% 
  mutate_each(funs(normalize_table), -OTUID) %>% 
  gather(R2_222:RF_343, key = sample, value = abundance) %>% 
  group_by(OTUID) %>% 
  filter(any(abundance > 0.01)) %>% 
  ungroup() %>% 
  spread(key = sample, value = abundance) %>% 
  mutate(OTUID = as.integer(OTUID)) %>% 
  inner_join(ramp_23_lefse, by = "OTUID") %>% 
  arrange(breaks, desc(LDA))

scalewhiteblack <- colorRampPalette(c("white", "black"), space = "rgb")(100)
lmat = rbind(c(4,0,0),c(2,1,3))
lwid = c(0.60,1.9,0.3)
lhei = c(0.3,1.5)

pdf("Figure3_panelC.pdf", width = 12, height = 9)
ramp_23_otu_lefse %>% 
  select(starts_with("R", ignore.case = F)) %>% 
  t() %>% 
  heatmap.2(Rowv = F, Colv = F, col = scalewhiteblack, margins = c(13, 9.5), trace = "none", density.info = "none", xlab = "", ylab = "", main = "", srtCol = 67.5, cexCol = 1.3, cexRow = 2.0, lmat = lmat, lwid = lwid, lhei = lhei, labCol = ramp_23_otu_lefse$family)
dev.off()
```




